File: linkedin-optimizer/lib/analyzer.ts
const OLLAMA_API = 'http://localhost:11434/api/generate';
const MODEL = 'qwen3:8b';

export async function analyzeProfile(data: any) {
  console.log('🔍 Starting profile analysis...');
  console.log('📊 Profile data received:', {
    hasProfile: !!data.profile,
    hasPosts: !!data.posts,
    postsCount: data.posts?.length || 0
  });
  
  const systemPrompt = `You are an expert LinkedIn profile optimization consultant. Analyze profiles AND posts to provide SPECIFIC, ACTIONABLE insights.

CRITICAL RULES:
1. Be SPECIFIC - Quote actual data, give exact numbers
2. Analyze BOTH profile AND posts
3. Identify EXACT problems with examples
4. Give CONCRETE action items
5. Return ONLY valid JSON

OUTPUT FORMAT:
{
  "score": <0-100>,
  "scoreDescription": "<one sentence>",
  "dataQualityIssues": [{"type": "error", "title": "<issue>", "description": "<details>"}],
  "optimizations": [{"type": "warning", "title": "<suggestion>", "description": "<action>"}],
  "strengths": [{"type": "success", "title": "<strength>", "description": "<why>"}],
  "postAnalysis": [{"type": "info", "title": "<post insight>", "description": "<improvement>"}],
  "priorityActions": ["<action 1>", "<action 2>", "<action 3>"]
}`;

  const userPrompt = `Analyze this LinkedIn profile INCLUDING posts:

${JSON.stringify(data, null, 2)}

Check:
1. Data quality (junk data, UI elements)
2. Profile completeness
3. Content quality (metrics, keywords)
4. Post quality (hooks, structure, hashtags, CTAs)
5. Engagement potential

For posts, analyze:
- Content quality and storytelling
- Structure and formatting
- Hashtag usage
- Call-to-action
- Specific improvements with examples

Return ONLY JSON.`;

  try {
    console.log('🌐 Attempting to connect to Ollama...');
    
    // Add timeout to fetch
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 3000); // 3 second timeout
    
    const response = await fetch(OLLAMA_API, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: MODEL,
        prompt: systemPrompt + "\n\n" + userPrompt,
        stream: false,
        options: {
          temperature: 0.3,
          num_predict: 3000,
          top_p: 0.9,
          top_k: 40
        }
      }),
      signal: controller.signal
    }).catch((err) => {
      clearTimeout(timeoutId);
      console.log('❌ Ollama connection failed:', err.message);
      throw new Error('Ollama not available');
    });

    clearTimeout(timeoutId);
    console.log('✅ Ollama responded');

    if (!response.ok) {
      throw new Error(`Ollama error: ${response.statusText}`);
    }

    const result = await response.json();
    const responseText = result.response;
    
    // Extract JSON
    const jsonMatch = responseText.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
    
    throw new Error('No JSON in response');
  } catch (error) {
    // Silently use fallback - no console error
    console.log('⚠️ Ollama not available, using fallback analysis');
    return generateFallbackInsights(data);
  }
}

function generateFallbackInsights(data: any) {
  console.log('🔄 Generating fallback insights...');
  const profile = data.profile || {};
  
  // Handle both data structures: profile.basic.name OR profile.name
  const basic = profile.basic || profile;
  const name = basic.name || profile.name || '';
  const location = basic.location || profile.location || '';
  const connections = basic.connections || profile.connections || '0';
  const followers = basic.followers || profile.followers || '0';
  
  const about = profile.about || '';
  const experience = profile.experience || [];
  const education = profile.education || [];
  const skills = profile.skills || [];
  const certifications = profile.certifications || [];
  const posts = data.posts || [];
  const analytics = data.analytics || {};
  
  const issues = [];
  const optimizations = [];
  const strengths = [];
  const postAnalysis = [];
  const actions = [];
  
  let score = 50;

  // Clean junk data from education (UI elements scraped by mistake)
  const junkKeywords = ['Report this', 'Copy', 'LinkedIn', 'Facebook', 'X', 'Posts', 'Comments', 'Reactions'];
  const cleanEducation = education.filter((edu: any) => {
    const schoolName = (edu.school || '').toLowerCase();
    return !junkKeywords.some(keyword => schoolName.includes(keyword.toLowerCase()));
  });

  // Profile basic info analysis
  if (!name || name.trim() === '') {
    issues.push({
      type: 'error',
      title: 'Missing Name',
      description: 'Profile name is empty. Add your full professional name.'
    });
    score -= 20;
  } else {
    strengths.push({
      type: 'success',
      title: 'Profile Name Present',
      description: `"${name}" is clear and professional. Good foundation for your profile.`
    });
    score += 5;
  }

  // Location check
  if (!location || location.trim() === '') {
    optimizations.push({
      type: 'warning',
      title: 'Missing Location',
      description: 'Add your location (city, country) to help recruiters find you and show up in local searches.'
    });
    actions.push('Add your location to improve discoverability');
    score -= 5;
  } else {
    strengths.push({
      type: 'success',
      title: 'Location Added',
      description: `Location helps with local job searches.`
    });
    score += 3;
  }

  // Connections analysis
  const connectionsNum = connections === '500+' ? 500 : parseInt(connections) || 0;
  if (connectionsNum < 100) {
    optimizations.push({
      type: 'warning',
      title: `Low Connection Count: ${connections}`,
      description: 'Target 500+ connections. Connect with colleagues, classmates, industry peers, and recruiters. Send 10 personalized connection requests daily.'
    });
    actions.push(`Grow network from ${connections} to 500+: connect with colleagues, alumni, and industry professionals`);
    score -= 10;
  } else if (connectionsNum < 500) {
    optimizations.push({
      type: 'warning',
      title: `${connections} Connections - Keep Growing`,
      description: 'You\'re on the right track! Aim for 500+ to unlock "500+" badge and increase visibility.'
    });
    score += 5;
  } else {
    strengths.push({
      type: 'success',
      title: `Strong Network: ${connections} Connections`,
      description: 'Excellent network size! Continue engaging with your connections regularly.'
    });
    score += 10;
  }

  // About section
  const aboutLength = about.length;
  if (aboutLength === 0) {
    issues.push({
      type: 'error',
      title: 'About Section Missing',
      description: 'Current: 0 characters. Target: 1000-2000 characters. Write 3 paragraphs: (1) Who you are + years of experience, (2) Key achievements with numbers, (3) What you\'re looking for.'
    });
    actions.push('Write 1000+ character About section with your story, achievements, and career goals');
    score -= 15;
  } else if (aboutLength < 500) {
    optimizations.push({
      type: 'warning',
      title: `About Section Too Short: ${aboutLength} characters`,
      description: 'Expand to 1000-2000 characters. Add specific achievements, skills, and what makes you unique.'
    });
    actions.push('Expand About section with specific achievements and metrics');
    score -= 5;
  } else {
    strengths.push({
      type: 'success',
      title: `About Section: ${aboutLength} characters`,
      description: 'Good length! Make sure it includes specific achievements and keywords.'
    });
    score += 10;
  }

  // Experience analysis
  const experienceCount = experience.length;
  if (experienceCount === 0) {
    issues.push({
      type: 'error',
      title: 'No Work Experience Listed',
      description: 'Add your work history with 3-5 bullet points per role. Include metrics (%, $, time saved) to show impact.'
    });
    actions.push('Add work experience with quantifiable achievements');
    score -= 15;
  } else {
    strengths.push({
      type: 'success',
      title: `${experienceCount} Work Experience${experienceCount > 1 ? 's' : ''} Listed`,
      description: 'Good career history. Ensure each role has 3-5 bullet points with metrics (%, $, time saved, etc.).'
    });
    score += 10;
  }

  // Education analysis
  const educationCount = cleanEducation.length;
  if (educationCount === 0) {
    optimizations.push({
      type: 'warning',
      title: 'No Education Listed',
      description: 'Add your educational background including degrees, certifications, and relevant coursework.'
    });
    actions.push('Add educational qualifications and certifications');
    score -= 10;
  } else {
    // Don't show junk data warning, just use clean count
    strengths.push({
      type: 'success',
      title: `${educationCount} Education Entr${educationCount > 1 ? 'ies' : 'y'}`,
      description: 'Educational background is documented.'
    });
    score += 5;
  }

  // Skills analysis
  const skillsCount = skills.length;
  if (skillsCount === 0) {
    issues.push({
      type: 'error',
      title: 'No Skills Listed',
      description: 'Add 15+ skills including: (1) Technical skills (languages, tools), (2) Soft skills (leadership, communication), (3) Industry-specific skills. Get endorsements from colleagues.'
    });
    actions.push('Add 15+ relevant skills and request endorsements');
    score -= 15;
  } else if (skillsCount < 15) {
    optimizations.push({
      type: 'warning',
      title: `Only ${skillsCount} Skill${skillsCount > 1 ? 's' : ''} Listed`,
      description: `Add ${15 - skillsCount} more skills to reach 15+. Include technical skills, soft skills, and industry-specific expertise.`
    });
    actions.push(`Add ${15 - skillsCount} more skills to reach 15+ total`);
    score -= 5;
  } else {
    strengths.push({
      type: 'success',
      title: `${skillsCount} Skills Listed`,
      description: 'Good skill coverage! Make sure to get endorsements from colleagues.'
    });
    score += 10;
  }

  // Certifications
  const certsCount = certifications.length;
  if (certsCount === 0) {
    optimizations.push({
      type: 'warning',
      title: 'No Certifications',
      description: 'Add relevant certifications, courses, or training to demonstrate continuous learning.'
    });
    score -= 5;
  } else {
    strengths.push({
      type: 'success',
      title: `${certsCount} Certification${certsCount > 1 ? 's' : ''}`,
      description: 'Shows commitment to professional development.'
    });
    score += 5;
  }

  // Post analysis - Individual post insights
  // Handle both data structures: posts array with 'content' field OR experience array with 'description' field
  let validPosts = posts.filter((p: any) => p.content && p.content.length > 50 && !p.content.includes('Report this'));
  
  // If no valid posts in posts array, check experience array (scraper sometimes puts posts there)
  if (validPosts.length === 0 && experience.length > 0) {
    validPosts = experience
      .filter((exp: any) => exp.description && exp.description.length > 100 && !exp.description.startsWith('Report this'))
      .map((exp: any) => ({
        content: exp.description,
        likes: 0,
        comments: 0,
        shares: 0,
        hasHashtags: exp.description.includes('#'),
        wordCount: exp.description.split(/\s+/).length
      }));
  }
  
  if (validPosts.length === 0) {
    optimizations.push({
      type: 'warning',
      title: 'No Posts Found',
      description: 'Start posting 2-3 times per week. Share insights, achievements, and industry trends to increase visibility.'
    });
    postAnalysis.push({
      type: 'warning',
      title: 'No Content Activity',
      description: 'Create a content calendar: post industry insights, personal achievements, and tips weekly. Aim for 150-300 words per post with 3-5 hashtags.'
    });
    actions.push('Start posting 2-3 times per week with industry insights and achievements');
    score -= 10;
  } else {
    // Analyze each post
    postAnalysis.push({
      type: 'info',
      title: `${validPosts.length} Post${validPosts.length > 1 ? 's' : ''} Analyzed`,
      description: 'Here are specific improvements for your posts:'
    });
    score += 10;

    validPosts.slice(0, 5).forEach((post: any, index: number) => {
      const content = post.content || post.description || '';
      const wordCount = content.split(/\s+/).length;
      const hasHashtags = post.hasHashtags || content.includes('#');
      const engagement = (post.likes || 0) + (post.comments || 0);
      
      // Extract first 100 chars for preview
      const preview = content.substring(0, 100).replace(/\s+/g, ' ').trim();
      
      const postIssues = [];
      if (wordCount < 100) postIssues.push(`too short (${wordCount} words, aim for 150-300)`);
      if (!hasHashtags) postIssues.push('missing hashtags');
      if (engagement < 10) postIssues.push('low engagement');
      
      if (postIssues.length > 0) {
        postAnalysis.push({
          type: 'warning',
          title: `Post #${index + 1}: ${postIssues.join(', ')}`,
          description: `"${preview}..." - Improve by: ${!hasHashtags ? 'Add 3-5 relevant hashtags. ' : ''}${wordCount < 100 ? 'Expand content with storytelling and examples. ' : ''}${engagement < 10 ? 'Ask questions to boost engagement.' : ''}`
        });
      } else {
        postAnalysis.push({
          type: 'success',
          title: `Post #${index + 1}: Good quality`,
          description: `"${preview}..." - ${wordCount} words, ${hasHashtags ? 'has hashtags' : 'no hashtags'}, ${engagement} engagements. Keep it up!`
        });
      }
    });

    // Overall post insights
    const avgWordCount = validPosts.reduce((sum: number, p: any) => {
      const content = p.content || p.description || '';
      return sum + content.split(/\s+/).length;
    }, 0) / validPosts.length;
    const postsWithHashtags = validPosts.filter((p: any) => {
      const content = p.content || p.description || '';
      return p.hasHashtags || content.includes('#');
    }).length;
    
    if (postsWithHashtags < validPosts.length / 2) {
      postAnalysis.push({
        type: 'warning',
        title: 'Missing Hashtags in Most Posts',
        description: `Only ${postsWithHashtags}/${validPosts.length} posts have hashtags. Add 3-5 relevant hashtags to each post for better reach. Example: #Industry #Skill #Topic`
      });
    }

    if (avgWordCount < 100) {
      postAnalysis.push({
        type: 'warning',
        title: 'Posts Too Short Overall',
        description: `Average ${Math.round(avgWordCount)} words. Aim for 150-300 words for better engagement and storytelling.`
      });
    }
  }

  // Analytics-based insights
  if (analytics.profileCompleteness !== undefined) {
    if (analytics.profileCompleteness < 50) {
      issues.push({
        type: 'error',
        title: `Profile Only ${analytics.profileCompleteness}% Complete`,
        description: 'Complete profiles get 21x more views. Fill in all sections: About, Experience, Education, Skills, and Certifications.'
      });
    } else if (analytics.profileCompleteness < 80) {
      optimizations.push({
        type: 'warning',
        title: `Profile ${analytics.profileCompleteness}% Complete`,
        description: 'Almost there! Complete remaining sections to reach 100% and maximize visibility.'
      });
    } else {
      strengths.push({
        type: 'success',
        title: `Profile ${analytics.profileCompleteness}% Complete`,
        description: 'Excellent profile completeness! This significantly boosts your visibility.'
      });
    }
  }

  // Ensure we have at least 3 priority actions
  if (actions.length === 0) {
    actions.push('Complete all profile sections (About, Experience, Education, Skills)');
    actions.push('Grow your network to 500+ connections');
    actions.push('Start posting valuable content 2-3 times per week');
  }

  const result = {
    score: Math.max(0, Math.min(100, score)),
    scoreDescription: score >= 80 ? 'Excellent profile with minor improvements needed' : 
                      score >= 60 ? 'Good profile with room for optimization' : 
                      score >= 40 ? 'Profile needs significant improvements' :
                      'Profile requires major optimization',
    dataQualityIssues: issues,
    optimizations: optimizations,
    strengths: strengths,
    postAnalysis: postAnalysis.length > 0 ? postAnalysis : [{
      type: 'info',
      title: 'Content Strategy Needed',
      description: 'Build a content strategy: Share industry insights, personal achievements, lessons learned, and tips. Post 2-3 times per week.'
    }],
    priorityActions: actions.slice(0, 5)
  };
  
  console.log('✅ Analysis complete! Score:', result.score);
  return result;
}

File: backend/analyzer.py
from typing import Dict, List, Any
import re

class ProfileAnalyzer:
    def __init__(self):
        self.weights = {
            'headline': 15,
            'about': 20,
            'experience': 25,
            'skills': 20,
            'education': 15,
            'completeness': 5
        }
        
        # Keywords for different roles
        self.power_words = ['expert', 'specialist', 'lead', 'senior', 'manager', 'director', 
                           'architect', 'engineer', 'developer', 'designer', 'consultant']
        self.action_verbs = ['built', 'created', 'led', 'managed', 'developed', 'designed',
                            'implemented', 'launched', 'achieved', 'increased', 'improved']
    
    def analyze(self, profile: Dict[str, Any]) -> Dict[str, Any]:
        """Deep analysis of LinkedIn profile including posts and achievements"""
        
        scores = {}
        insights = []
        recommendations = []
        strengths = []
        weaknesses = []
        
        name = profile.get('name', '')
        headline = profile.get('headline', '')
        about = profile.get('about', '')
        aboutAnalysis = profile.get('aboutAnalysis', {})
        experience = profile.get('experience', [])
        skills = profile.get('skills', [])
        education = profile.get('education', [])
        certifications = profile.get('certifications', [])
        connections = profile.get('connections', '0')
        followers = profile.get('followers', '0')
        posts = profile.get('posts', [])
        activity = profile.get('activity', {})
        
        # Network strength analysis
        conn_num = int(''.join(filter(str.isdigit, str(connections))) or 0)
        if conn_num >= 500:
            strengths.append(f"Strong network: {connections} connections")
            insights.append({
                "type": "good",
                "message": f"✓ Excellent network size: {connections} connections"
            })
        elif conn_num >= 100:
            insights.append({
                "type": "info",
                "message": f"💡 Good network ({connections}). Keep connecting with industry peers"
            })
        else:
            insights.append({
                "type": "warning",
                "message": f"⚠ Build your network (currently {connections}). Aim for 500+"
            })
            recommendations.append("Connect with colleagues, classmates, and industry professionals")
        
        # Analyze headline
        headline_score = self._analyze_headline(headline)
        scores['headline'] = headline_score
        
        if headline_score >= 80:
            strengths.append("Strong headline with clear positioning")
            insights.append({
                "type": "good",
                "message": "✓ Headline is compelling and keyword-optimized"
            })
        elif headline_score >= 50:
            insights.append({
                "type": "warning",
                "message": f"⚠ Headline is {len(headline)} chars. Add your specialty and value (e.g., 'Senior Developer | Python Expert | Building Scalable Systems')"
            })
            recommendations.append("Use format: [Role] | [Expertise] | [Value Proposition]")
        else:
            weaknesses.append("Weak or missing headline")
            insights.append({
                "type": "warning",
                "message": "⚠ Create a powerful headline with keywords recruiters search for"
            })
            recommendations.append("Add headline with your role, skills, and what you do")
        
        # Check for power words in headline
        has_power_words = any(word in headline.lower() for word in self.power_words)
        if not has_power_words and headline:
            insights.append({
                "type": "info",
                "message": "💡 Add power words like 'Expert', 'Specialist', 'Lead' to your headline"
            })
        
        # Analyze about section
        about_score = self._analyze_about(about)
        scores['about'] = about_score
        about_words = len(about.split()) if about else 0
        
        if about_words == 0:
            weaknesses.append("Missing About section")
            insights.append({
                "type": "warning",
                "message": "⚠ Add an About section (200-300 words) telling your story"
            })
            recommendations.append("Write about: Who you are, what you do, your achievements, what you're looking for")
        elif about_words < 100:
            insights.append({
                "type": "warning",
                "message": f"⚠ About section is only {about_words} words. Expand to 200-300 words"
            })
            recommendations.append("Add: Your unique value, key achievements, career goals")
        elif about_words >= 200:
            strengths.append("Comprehensive About section")
            insights.append({
                "type": "good",
                "message": f"✓ About section is detailed ({about_words} words)"
            })
        
        # Check for action verbs in about
        if about:
            has_action_verbs = any(verb in about.lower() for verb in self.action_verbs)
            if not has_action_verbs:
                insights.append({
                    "type": "info",
                    "message": "💡 Use action verbs: 'Built', 'Led', 'Achieved', 'Increased' to show impact"
                })
        
        # Analyze experience
        exp_score = self._analyze_experience(experience)
        scores['experience'] = exp_score
        
        if len(experience) >= 3:
            strengths.append(f"{len(experience)} work experiences listed")
            insights.append({
                "type": "good",
                "message": f"✓ {len(experience)} positions show career growth"
            })
        elif len(experience) > 0:
            insights.append({
                "type": "info",
                "message": f"💡 Add more past roles (currently {len(experience)}). Include internships, freelance work"
            })
        else:
            weaknesses.append("No experience listed")
            insights.append({
                "type": "warning",
                "message": "⚠ Add your work experience with bullet points showing achievements"
            })
        
        # Check experience descriptions
        if experience and len(experience) > 0:
            first_exp = experience[0]
            if 'title' in first_exp and len(str(first_exp.get('title', ''))) < 10:
                insights.append({
                    "type": "info",
                    "message": "💡 Add detailed descriptions to each role with quantifiable achievements"
                })
        
        # Analyze skills
        skill_score = self._analyze_skills(skills)
        scores['skills'] = skill_score
        
        if len(skills) == 0:
            weaknesses.append("No skills listed")
            insights.append({
                "type": "warning",
                "message": "⚠ Add 15-20 relevant skills. This is crucial for recruiter searches!"
            })
            recommendations.append("Add technical skills, soft skills, tools, and certifications")
        elif len(skills) < 10:
            insights.append({
                "type": "warning",
                "message": f"⚠ Only {len(skills)} skills listed. Add {15 - len(skills)} more to reach 15+"
            })
            recommendations.append("Add: Industry-specific skills, software tools, methodologies")
        elif len(skills) >= 15:
            strengths.append(f"{len(skills)} skills showcased")
            insights.append({
                "type": "good",
                "message": f"✓ Great! {len(skills)} skills make you discoverable in searches"
            })
        
        # Analyze education
        edu_score = self._analyze_education(education)
        scores['education'] = edu_score
        
        if len(education) > 0:
            strengths.append("Education credentials added")
        else:
            insights.append({
                "type": "info",
                "message": "💡 Add your education (degree, certifications, courses)"
            })
        
        # Profile completeness check
        completeness_items = []
        if name: completeness_items.append('name')
        if headline: completeness_items.append('headline')
        if about: completeness_items.append('about')
        if experience: completeness_items.append('experience')
        if skills: completeness_items.append('skills')
        if education: completeness_items.append('education')
        
        completeness = (len(completeness_items) / 6) * 100
        
        if completeness < 100:
            missing = []
            if not headline: missing.append('headline')
            if not about: missing.append('about section')
            if not experience: missing.append('experience')
            if not skills: missing.append('skills')
            if not education: missing.append('education')
            
            insights.append({
                "type": "info",
                "message": f"📋 Profile is {int(completeness)}% complete. Add: {', '.join(missing)}"
            })
        
        # Calculate overall score
        total_score = sum(scores[key] * self.weights[key] / 100 for key in scores)
        
        # Priority recommendations
        if not headline or headline_score < 50:
            recommendations.insert(0, "🎯 PRIORITY: Create a keyword-rich headline")
        if about_words < 100:
            recommendations.insert(0 if not recommendations else 1, "🎯 PRIORITY: Write a compelling About section")
        if len(skills) < 10:
            recommendations.insert(0 if not recommendations else 2, "🎯 PRIORITY: Add at least 15 skills")
        
        return {
            "score": int(total_score),
            "insights": insights[:8],  # Top 8 insights
            "recommendations": recommendations[:5],  # Top 5 recommendations
            "strengths": strengths,
            "weaknesses": weaknesses
        }
    
    def _analyze_headline(self, headline: str) -> int:
        if not headline:
            return 0
        score = 50
        if len(headline) > 30:
            score += 20
        if any(word in headline.lower() for word in ['expert', 'specialist', 'engineer', 'manager', 'developer']):
            score += 15
        if '|' in headline or '•' in headline:
            score += 15
        return min(score, 100)
    
    def _analyze_about(self, about: str) -> int:
        if not about:
            return 0
        words = len(about.split())
        if words < 50:
            return 30
        elif words < 100:
            return 50
        elif words < 200:
            return 75
        else:
            return 95
    
    def _analyze_experience(self, experience: List[Dict]) -> int:
        if not experience:
            return 0
        score = min(len(experience) * 25, 100)
        return score
    
    def _analyze_skills(self, skills: List[str]) -> int:
        if not skills:
            return 0
        score = min(len(skills) * 5, 100)
        return score
    
    def _analyze_education(self, education: List[Dict]) -> int:
        if not education:
            return 50
        return min(len(education) * 50, 100)

        
        # About section deep analysis
        if aboutAnalysis:
            if aboutAnalysis.get('hasActionVerbs'):
                strengths.append("About section uses action verbs")
            else:
                insights.append({
                    "type": "info",
                    "message": "💡 Add action verbs to About: 'Built', 'Led', 'Achieved', 'Delivered'"
                })
            
            if aboutAnalysis.get('hasNumbers') or aboutAnalysis.get('hasPercentages'):
                strengths.append("About section includes quantifiable achievements")
                insights.append({
                    "type": "good",
                    "message": "✓ Great! Your About section has numbers/metrics"
                })
            else:
                insights.append({
                    "type": "warning",
                    "message": "⚠ Add numbers to About: '30% increase', '5 years', '$1M revenue'"
                })
        
        # Experience deep analysis
        if experience:
            exp_with_achievements = sum(1 for exp in experience if exp.get('hasQuantifiableResults'))
            exp_with_bullets = sum(1 for exp in experience if exp.get('bulletPoints', 0) > 0)
            
            if exp_with_achievements > 0:
                strengths.append(f"{exp_with_achievements} roles with quantifiable achievements")
                insights.append({
                    "type": "good",
                    "message": f"✓ {exp_with_achievements} roles show measurable impact"
                })
            else:
                insights.append({
                    "type": "warning",
                    "message": "⚠ Add numbers to experience: 'Increased sales by 40%', 'Led team of 10'"
                })
                recommendations.append("Add quantifiable achievements to each role")
            
            # Check for detailed descriptions
            short_descriptions = sum(1 for exp in experience if exp.get('descriptionLength', 0) < 100)
            if short_descriptions > len(experience) / 2:
                insights.append({
                    "type": "info",
                    "message": f"💡 {short_descriptions} roles need detailed descriptions with bullet points"
                })
        
        # Skills analysis with endorsements
        if isinstance(skills, list) and len(skills) > 0:
            if isinstance(skills[0], dict):
                endorsed_skills = [s for s in skills if s.get('endorsements', 0) > 0]
                if endorsed_skills:
                    avg_endorsements = sum(s.get('endorsements', 0) for s in endorsed_skills) / len(endorsed_skills)
                    insights.append({
                        "type": "good",
                        "message": f"✓ Skills have avg {int(avg_endorsements)} endorsements"
                    })
                else:
                    insights.append({
                        "type": "info",
                        "message": "💡 Ask colleagues to endorse your top skills"
                    })
        
        # Certifications analysis
        if certifications and len(certifications) > 0:
            strengths.append(f"{len(certifications)} certifications listed")
            insights.append({
                "type": "good",
                "message": f"✓ {len(certifications)} certifications boost credibility"
            })
        else:
            insights.append({
                "type": "info",
                "message": "💡 Add certifications (Coursera, Udemy, industry certs)"
            })
        
        # Posts/Activity analysis
        if posts and len(posts) > 0:
            avg_engagement = activity.get('avgLikes', 0) + activity.get('avgComments', 0)
            
            if avg_engagement > 50:
                strengths.append("High post engagement")
                insights.append({
                    "type": "good",
                    "message": f"✓ Excellent engagement: avg {avg_engagement} reactions per post"
                })
            elif avg_engagement > 10:
                insights.append({
                    "type": "info",
                    "message": f"💡 Good engagement ({avg_engagement}/post). Post more consistently"
                })
            else:
                insights.append({
                    "type": "warning",
                    "message": "⚠ Low post engagement. Try: industry insights, personal stories, tips"
                })
            
            # Content quality analysis
            posts_with_hashtags = activity.get('postsWithHashtags', 0)
            posts_with_media = activity.get('postsWithMedia', 0)
            
            if posts_with_hashtags < len(posts) / 2:
                insights.append({
                    "type": "info",
                    "message": "💡 Use 3-5 relevant hashtags per post for better reach"
                })
            
            if posts_with_media < len(posts) / 2:
                insights.append({
                    "type": "info",
                    "message": "💡 Posts with images/videos get 2x more engagement"
                })
                recommendations.append("Add visuals to your posts (images, infographics, videos)")
        else:
            insights.append({
                "type": "warning",
                "message": "⚠ No recent posts found. Share content weekly to stay visible"
            })
            recommendations.append("Post 2-3 times per week: industry insights, achievements, tips")
        
        # Calculate comprehensive score
        total_score = (
            headline_score * 0.10 +
            about_score * 0.15 +
            exp_score * 0.20 +
            skill_score * 0.15 +
            edu_score * 0.10 +
            (len(certifications) * 5) +
            (min(conn_num / 10, 15)) +
            (min(len(posts) * 2, 10))
        )
        
        return {
            "score": int(min(total_score, 100)),
            "insights": insights[:10],
            "recommendations": recommendations[:6],
            "strengths": strengths,
            "weaknesses": weaknesses,
            "detailedMetrics": {
                "connections": connections,
                "followers": followers,
                "postsAnalyzed": len(posts),
                "avgEngagement": activity.get('avgLikes', 0) + activity.get('avgComments', 0),
                "certificationsCount": len(certifications),
                "skillsCount": len(skills),
                "experienceCount": len(experience)
            }
        }
